File: go/MyBot.go
package main

import (
	"os"
)

type MyBot struct {
	goalQueue *SearchQueue
}

//NewBot creates a new instance of your bot
func NewBot(s *State) Bot {
	mb := new(MyBot)
	mb.goalQueue = NewSearchQueue()
	s.ObservedSquares = make(SquareSet)
	s.LivingAnts = make(map[int]*Ant)

	Log.Printf("New bot created!")

	return mb
}

//DoTurn is where you should do your bot's actual work.
func (mb *MyBot) DoTurn(s *State) os.Error {
	Log.Printf("BFS: Search queue has size %v (from previous turns)", mb.goalQueue.Len())

	// Update map visibility
	Log.Printf("Updating visiblity for %v ants", len(s.LivingAnts))
	updated := 0
	for _, ant := range s.LivingAnts {
		updated += ant.square.Visit(s)
	}
	Log.Printf("Updated visiblity of %v squares", updated)

	// restore any newly visible squares to the goalQueue if they were previously set aside
	restoredSearchNodes := 0
	// TODO have an index of newly observed squares to avoid walking all?
	for _, square := range s.ObservedSquares {
		for _, searchNode := range square.deferredSearchNodes {
			mb.goalQueue.Push(searchNode)
			restoredSearchNodes++
		}

		square.deferredSearchNodes = make([]*SearchNode, 0)
	}
	Log.Printf("BFS: Restored %v deferred search nodes from previous turns", restoredSearchNodes)

	// Compute game statistics for weighting model
	s.Stats.Update(s)
	Log.Printf("Current turn statistics are %+v", s.Stats)

	// Make a shared list of goals used by all ants
	// TODO can skip this until we actually need to pick a goal
	Log.Printf("Looking for goals")
	s.GenerateGoals()

	// Loop over all goals and seed them into the search queue if new, or clean them up if invalid
	for _, goal := range AllGoals {
		square := goal.Destination()
		if !square.HasGoal(goal) {
			// Goal is new
			route := make(Route, 0)
			newNode := NewSearchNode(square, goal, route)
			Log.Printf("BFS: Adding seed node: %+v", newNode)
			mb.goalQueue.Push(newNode)
		} else if !goal.IsValid() {
			// Goal should quiesce
			goal.Die()
		} // else do nothing to goal
	}

	Log.Printf("Search queue has size %v after goal generation", mb.goalQueue.Len())

	maxSearchRadius := 10 // hack to limit memory usage
	searchRadius := 0
	searchCount := 0

	searchTimeNanos := (int64)(s.TurnTime * 800000)
	RunTimeoutLoop(searchTimeNanos, func() bool {
		if mb.goalQueue.Len() == 0 {
			return false // stop looping
		}

		// visit the first node in the queue and unpack it
		node := mb.goalQueue.Pop()
		nodeSearchRadius := len(node.route)
		if nodeSearchRadius >= searchRadius {
			searchRadius = nodeSearchRadius
		} else {
			panic("search is not breadth-first!")
		}
		searchCount++

		//Log.Printf("BFS searching %+v", node)
		square, goal, route := node.square, node.goal, node.route

		// Purge from queue if no longer valid
		if !goal.IsValid() {
			return true // next iteration
		}

		// Record the route to this goal on the square
		square.goals[goal.Id()] = route

		// don't search any further from this node if we've maxed out
		if nodeSearchRadius >= maxSearchRadius {
			return true
		}

		// put neighboring squares at end of search queue
		for _, neighbor := range square.Neighbors() {
			//Log.Printf("BFS: looking for new search node at %v", neighbor)

			// Don't enqueue the neighbor if we've already visited it for this goal
			if neighbor.HasGoal(goal) {
				//Log.Printf("BFS: skipping already visited square %v", neighbor)
			} else {
				newRoute := make(Route, 0)
				newRoute = append(newRoute, square)
				newRoute = append(newRoute, route...)
				newNode := NewSearchNode(neighbor, goal, newRoute)

				// Don't try to search nodes we haven't observed yet (they could
				// be water). Instead, set aside those nodes and restore them
				// later
				if !neighbor.observed {
					//Log.Printf("BFS: skipping unobserved square %v", neighbor)
					neighbor.deferredSearchNodes = append(neighbor.deferredSearchNodes, newNode)
				} else {
					//Log.Printf("BFS: Adding new node: %+v", newNode)
					mb.goalQueue.Push(newNode)
				}
			}
		}

		return true // continue looping
	})

	// TODO restore the plug goal?
	Log.Printf("BFS: done searching. Search count was %v, radius was at most %v square from goals", searchCount, searchRadius)

	// Issue orders for each ant's best-available goal
	// TODO this should be treated as a queue, not an array
	for _, ant := range s.LivingAnts {
		// check passable squares
		square := ant.square
		passable := square.Neighbors().Minus(square.Blacklist())

		// try to assign a goal if we don't have one
		if ant.goal == nil {
			Log.Printf("Orders: finding new orders for %v", ant)

			// Iterate through all the square's goals and find the highest priority passable route
			var bestGoal Goal = nil
			for goalId, route := range square.goals {
				passableRoute := (len(route) == 0) || passable.Member(route[0])
				goal := goalId.Goal()
				if passableRoute && (bestGoal == nil || goal.Priority() > bestGoal.Priority()) {
					// TODO break priority ties by route length
					bestGoal = goal
				}
			}

			ant.SetGoal(bestGoal)
		}

		// Execute either the assigned route or a random one
		var route Route = nil
		if ant.goal == nil {
			route = PickWanderForAnt(s, ant)
		} else {
			route = ant.Route()
		}

		Log.Printf("Orders: route for %v is %v", ant, route)
		passableRoute := (len(route) == 0) || passable.Member(route[0])
		if len(route) > 0 && passableRoute {
			ant.OrderTo(s, route[0])
		} else {
			Log.Printf("Route is impassable, doing nothing")
		}
	}

	//returning an error will halt the whole program!
	return nil
}

File: ruby/MyBot.rb
$:.unshift File.dirname(__FILE__)
%w(ai ant containers items stats params_matrix goals log priority_queue search_node square timeout_loop).each { |lib| require lib }

module Enumerable
  def rand
    sort_by { Kernel.rand }.first
  end
end

AI.instance.setup do |ai|
  log "Mybot: Setup"
end

goal_search_queue = Containers::PriorityQueue.new

AI.instance.run do |ai|
  log "Search queue has size #{goal_search_queue.size} (from previous turns)"

  ants_to_move = Ant.living.dup

  # Update map visibility
  log "Updating visible squares for #{ants_to_move.count} ants"
  updated = ants_to_move.inject(0) { |total, ant| total + ant.square.visit! }
  log "Updated visibility of #{updated} squares"

  # TODO restore any newly visible squares to the goal_search_queue if
  # they were previously set aside

  # Compute game statistics for weighting model
  stats = Stats.new(ai)
  Goal.stats = stats
  log "Current turn statistics are #{stats.inspect}"

  # Make a shared list of goals used by all ants
  # TODO can skip this until we actually need to pick a goal
  log "Looking for goals"
  goals = Goal.all
  goal_stats = goals.group_by(&:class).map { |k, v| [k, v.size] }
  log "Found #{goals.size} initial goals - #{goal_stats.inspect}"

  # Purge all invalid ant goals
  # TODO can push this down to the second loop
  ants_to_move.each do |ant|
    if ant.goal && !ant.goal.valid?
      ant.goal = nil
    end
  end

  # Figure out which goals are new and seed them into the DFS queue
  goals.each do |goal|
    square = goal.square
    if !square.goals.has_key?(goal)
      goal_search_queue.push(SearchNode.new(goal.square, goal, []), 2)
    end
  end

  log "Search queue has size #{goal_search_queue.size} after goal generation"

  search_radius = 0; search_count = 0 # instrument how far we were able to search
  TimeoutLoop.run((AI.instance.turntime / 1000.0) * 0.5) do
    # visit the first node in the queue and unpack it
    node = goal_search_queue.pop
    if node.nil?
      log "BFS: No more squares to search"
      TimeoutLoop.halt!
      next
    end

    search_radius = node.route.size
    search_count += 1

    square = node.square
    goal = node.goal
    route = node.route

    # Purge from queue if no longer valid
    next unless goal.valid?

    # Record the route to this goal on the square
    square.goals[goal] = route

    # put neighboring squares at end of search queue
    square.neighbors.each do |neighbor|
      # TODO instead of skipping, need to put this on a retry queue
      next if !neighbor.observed?

      # Don't enqueue the neighbor if we've already visited it for this goal
      next if neighbor.goals.has_key?(goal)

      new_route = [square] + route
      goal_search_queue.push(SearchNode.new(neighbor, goal, new_route), 1.0 / new_route.size)
    end
  end

  # TODO restore the plug goal?

  log "BFS: done searching. Search count was #{search_count}, radius was at least #{search_radius - 1} squares from goals"

  # Issue orders for each ant's best-available goal
  TimeoutLoop.run((AI.instance.turntime / 1000.0) * 0.2) do
    ant = ants_to_move.shift
    if ant.nil?
      log "Issued orders to all ants before timing out"
      TimeoutLoop.halt!
      next
    end

    log "Next ant in queue is #{ant}; after this we have #{ants_to_move.size} to move."

    # Find the best goal that this square knows about and is passable
    square = ant.square
    passable = square.neighbors - square.blacklist

    # Iterate through all the square's goals doing two things: purge invalids, and find highest priority
    best_goal = Wander.instance
    best_route = Wander.pick_route_for_ant(ant)

    log "Square has #{square.goals.size} goals attached"

    square.goals.each do |goal, route|
      if goal.valid?
        if goal.priority > best_goal.priority && passable.member?(route.first)
          best_goal = goal
          best_route = route
        end
      else
        square.goals.delete(goal)
      end
    end

    log "Best passable goal from #{square} is #{best_goal} with route #{best_route}"
    ant.goal = best_goal # required for escortability check

    if best_route.empty?
      log "#{ant} has reached destination (or is stuck) and will not move"
    else
      log "#{ant} will move to #{best_route.first}"
      ant.order_to best_route.first
    end
  end
end
